### ISSUE TITLE
$http doesn't immediately increment $brower's outstandingRequestCount. Protractor issues result.

### ISSUE THREAD

**Issue by wbyoko** [angular/angular.js#13782]
http://jsfiddle.net/awtwff3x/ (using Angular 1.4.8)

The issue is that using $http doesn't seem to register with $browser.defer synchronously so that $browser.notifyWhenNoOutstandingRequests waits accordingly.

This has created an issue where protractor tests are flaky because they aren't waiting for resources to load. 

Also seems with $timeout registering synchronously using a $timeout anywhere in a controller will allow for the $http request to register and have the protractor test wait correctly.

You can uncomment line 22 to see that requests queue after $timeout is used, otherwise notifyWhenNoOutstandingRequests returns synchronously.

Thanks.

**Comment by petebacondarwin** [comment]
I think this might be since we put the request through a `$q` promise that needs to be resolved async before the `$httpBackend` is triggered, right?

**Comment by wbyoko** [comment]
I have made pull requests for the 1.4 and master branches which fix this issue. 

The $q promise needing to be resolved async is the issue and `$browser.$$incOutstandingRequestCount()` needed to be called synchronously. Configuring this and `$browser.$$completeOutstandingRequest(noop)` to respectively be before and at the end of the promise chain yields the correct result.

Tests to confirm this interaction were added and when implemented locally protractor waited accordingly.
